### PR TITLE
Support writing timestamp with timezone in BigQuery

### PR DESCRIPTION
**Pull request by regadas**
## Description

Relates to https://github.com/trinodb/trino/issues/13741

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# BigQuery
* Add support for writing `timestamp with time zone` type. ({issue}`17793`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java [hunk 1]
```diff
@@ -103,6 +106,14 @@ public static Object readNativeValue(Type type, Block block, int position)
             int nanoAdjustment = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
             return DATETIME_FORMATTER.format(toZonedDateTime(epochSeconds, nanoAdjustment, UTC));
         }
+        if (type.equals(TIMESTAMP_TZ_MILLIS)) {
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java:109]: This branch isn't tested in my understanding.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 2]
```diff
@@ -264,8 +264,6 @@ protected Optional<DataMappingTestSetup> filterDataMappingSmokeTestData(DataMapp
             case "time(6)":
             case "timestamp":
             case "timestamp(3)":
-            case "timestamp(3) with time zone":
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:267]: I don't think the connector supports creating the table with this type. Please take a look at `BigQueryType#toStandardSqlTypeName` method.

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:267]: I'll have a better look at this. tbh I couldn't understand very well how this is being tested. Can you expand on this? What's the expectation here?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java [hunk 3]
```diff
@@ -160,18 +160,28 @@ public static String timeToStringConverter(Object value)
 
     @VisibleForTesting
     public static String timestampToStringConverter(Object value)
+    {
+        return timestampToStringConverter(value, DATETIME_FORMATTER);
```
[the remark this answers is not in the document]
Comment (regadas) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:164]: Can you expand on the signature for `convertToString` you are envisioning?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java (hunk 1 above)
Comment (regadas) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java:109]: Yeah this one doesnt make sense. removing

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (regadas) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java]: Never mind, I see what you are referring to here.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 2 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:267]: double checked and yeah this connector does not support the remove type.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java [hunk 4]
```diff
@@ -94,7 +94,7 @@ private BigQueryType() {}
             1, // 9 digits after the dot
     };
     private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("''HH:mm:ss.SSSSSS''");
-    private static final DateTimeFormatter DATETIME_FORMATTER = DateTimeFormatter.ofPattern("''yyyy-MM-dd HH:mm:ss.SSSSSS''");
+    private static final DateTimeFormatter DATETIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:97]: `yyyy` leads to incorrect result. Please reuse `BigQueryTypeUtils#DATETIME_FORMATTER`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 5]
```diff
@@ -562,10 +562,28 @@ public void testTimestampWithTimeZone()
...
         // TODO (https://github.com/trinodb/trino/pull/12210) Add support for timestamp with time zone type in views
     }
 
+    @Test
+    public void testUnsupportedTimestampWithTimeZone()
+    {
+        try (TestTable table = new TestTable(getBigQuerySqlExecutor(), "test.unsupported_tz", "(col timestamp)")) {
+            assertQueryFails("INSERT INTO " + table.getName() + " VALUES (timestamp '-0001-01-01 00:00:00.000000 UTC')", "Failed to insert rows.*");
+            assertQueryFails("INSERT INTO " + table.getName() + " VALUES (timestamp '0000-12-31 23:59:59.999999 UTC')", "Failed to insert rows.*");
+            assertQueryFails("INSERT INTO " + table.getName() + " VALUES (timestamp '10000-01-01 00:00:00.000000 UTC')", "Failed to insert rows.*");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:576]: Please add another case of valid timestamp with negative year. e.g. -2021-09-07 23:59:59.999999 UTC
The current logic doesn't throw an exception even though it's out of supported range.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 6]
```diff
@@ -562,10 +562,28 @@ public void testTimestampWithTimeZone()
                 // max value in BigQuery
                 .addRoundTrip("TIMESTAMP", "TIMESTAMP '9999-12-31 23:59:59.999999-00:00'",
                         TIMESTAMP_TZ_MICROS, "TIMESTAMP '9999-12-31 23:59:59.999999 UTC'")
+                .execute(getQueryRunner(), trinoCreateAsSelect("test.timestamp_tz"))
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:565]: Please test with several timezones. You can refer to `TestPostgreSqlTypeMapping#sessionZonesDataProvider`.

**Review by ebyhr — DISMISSED** [review]

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 4 above)
Comment (regadas) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:97]: I think for now I'll just fix that here and then do a follow up PR were we rely only on `BigQueryType#DATETIME_FORMATTER` and also update https://github.com/trinodb/trino/blob/f1f45bb38281c1565194528844880841132aa68d/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java#L103-L108

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java (hunk 6 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:565]: I guess here you are thinking in following the same pattern

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 7]
```diff
@@ -526,44 +532,96 @@ public void testTime()
...
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '2019-03-18 10:01:17.987000-07:31'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2019-03-18 17:32:17.987000 UTC'")
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '2021-09-07 23:59:59.999999-00:00'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2021-09-07 23:59:59.999999 UTC'")
-                // max value in BigQuery
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '9999-12-31 23:59:59.999999-00:00'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '9999-12-31 23:59:59.999999 UTC'")
+        SqlDataTypeTest test = SqlDataTypeTest.create();
+        for (String input : inputs) {
+            ZonedDateTime expected = LocalDateTime.parse(input, DATETIME_FORMATTER).atZone(zoneId).withZoneSameInstant(UTC);
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:596]: Please set a session `execute` method and avoid calculating the expected value on the fly. You can refer to `io.trino.plugin.postgresql.TestPostgreSqlTypeMapping#testTimestamp`.

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java]: I am not sure if I fully grasp the session magic 😅

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 8]
```diff
@@ -526,44 +529,89 @@ public void testTime()
                 .execute(getQueryRunner(), bigqueryViewCreateAndInsert("test.time"));
     }
 
-    @Test
-    public void testTimestampWithTimeZone()
+    @DataProvider
+    public Object[][] timestampsWithZoneDataProvider()
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:533]: Rename to sessionZonesDataProvider.

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:533]: Move to after `testTimestampWithTimeZone` method.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 9]
```diff
@@ -526,44 +529,89 @@ public void testTime()
...
+    {
+        return new Object[][] {
+                {UTC},
+                {ZoneId.of("America/Los_Angeles")},
+                {ZoneId.of("America/New_York")},
+                {ZoneId.of("Europe/London")},
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:539]: Why did you avoid using the same value of `TestPostgreSqlTypeMapping#sessionZonesDataProvider`?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 10]
```diff
@@ -526,44 +529,89 @@ public void testTime()
...
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '2019-03-18 10:01:17.987000+02:17'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2019-03-18 07:44:17.987000 UTC'")
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '2019-03-18 10:01:17.987000-07:31'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2019-03-18 17:32:17.987000 UTC'")
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '2021-09-07 23:59:59.999999-00:00'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2021-09-07 23:59:59.999999 UTC'")
-                // max value in BigQuery
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '9999-12-31 23:59:59.999999-00:00'",
-                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '9999-12-31 23:59:59.999999 UTC'")
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '0001-01-01 00:00:00.000'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '0001-01-01 00:00:00.000000 UTC'")
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:552]: Please revert the format change.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 11]
```diff
@@ -526,44 +529,89 @@ public void testTime()
...
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '1970-01-01 00:00:01.12345'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123450 UTC'")
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '1970-01-01 00:00:01.123456'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123456 UTC'")
+
+                // negative epoch
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '1969-12-31 23:59:59.999995'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '1969-12-31 23:59:59.999995 UTC'")
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '1969-12-31 23:59:59.999949'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '1969-12-31 23:59:59.999949 UTC'")
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '1969-12-31 23:59:59.999994'", TIMESTAMP_TZ_MICROS, "TIMESTAMP '1969-12-31 23:59:59.999994 UTC'")
+
+                .execute(getQueryRunner(), trinoCreateAsSelect("test.timestamp_tz"))
+                .execute(getQueryRunner(), trinoCreateAsSelect(session, "test.timestamp_tz"))
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:593]: Please add `.execute(getQueryRunner(), session, trinoCreateAsSelect("test.timestamp_tz"))`.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java (hunk 9 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:539]: no particular reason ... but now i know that you want the same values.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 12]
```diff
@@ -526,44 +533,122 @@ public void testTime()
                 .execute(getQueryRunner(), bigqueryViewCreateAndInsert("test.time"));
     }
 
-    @Test
-    public void testTimestampWithTimeZone()
+    @Test(dataProvider = "zonesDataProvider")
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:536]: The data provider name is wrong. Could you run this test before pushing a change?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 13]
```diff
@@ -526,44 +533,122 @@ public void testTime()
...
+    public void testTimestampWithTimeZone(ZoneId zoneId)
     {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey(zoneId.getId()))
+                .build();
+
         SqlDataTypeTest.create()
                 // min value in BigQuery
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '0001-01-01 00:00:00.000000 UTC'",
+                .addRoundTrip("TIMESTAMP", "TIMESTAMP '0001-01-01 00:00:00.000'",
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:545]: These tests are broken as far as I run locally.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java (hunk 12 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:536]: How do you run these tests locally?

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:536]: runnning `./mvnw clean install -B --strict-checksums -V -T 1C -pl ":trino-bigquery"` but don't seem to be included.

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:536]: There's another profile `cloud-tests`.

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:536]: totally missed that. Thanks

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java]: ok I believe I understand things a bit better. Sorry I've had limited time to spend on this. Thanks for the reviews and patience.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java (hunk 13 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java]: Test is passing now

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 14]
```diff
@@ -526,44 +533,128 @@ public void testTime()
...
+        testTimestampWithTimeZone("TIMESTAMP")
+                .execute(getQueryRunner(), bigqueryCreateAndInsert("test.timestamp_tz"));
+    }
+
+    private SqlDataTypeTest testTimestampWithTimeZone(String inputType)
+    {
+        return SqlDataTypeTest.create()
                 // min value in BigQuery
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '0001-01-01 00:00:00.000000 UTC'",
+                .addRoundTrip(inputType, "TIMESTAMP '0001-01-01 00:00:00.000 UTC'",
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:556]: It would be nice to order by old -> new date so that we can understand missing dates easily.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 15]
```diff
@@ -526,44 +533,128 @@ public void testTime()
...
+                .addRoundTrip(inputType, "TIMESTAMP '1970-01-01 00:13:42.000 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:13:42.000000 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '2018-04-01 02:13:55.123 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2018-04-01 02:13:55.123000 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '2018-03-25 03:17:17.000 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2018-03-25 03:17:17.000000 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '1986-01-01 00:13:07.000 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1986-01-01 00:13:07.000000 UTC'")
+
+                // same as above but with higher precision
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:580]: `2021-09-07 23:59:59.999999` doesn't exist in the above block. You can just remove this comment. I don't think it's helpful for readers.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 16]
```diff
@@ -526,44 +533,128 @@ public void testTime()
...
+                .addRoundTrip(inputType, "TIMESTAMP '1970-01-01 00:00:01.123 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123000 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '1970-01-01 00:00:01.1234 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123400 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '1970-01-01 00:00:01.12345 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123450 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '1970-01-01 00:00:01.123456 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:00:01.123456 UTC'")
+
+                // negative epoch
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:618]: What's the difference from "before epoch"? Please merge into one.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 17]
```diff
@@ -526,44 +533,128 @@ public void testTime()
...
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1970-01-01 00:13:42.123456 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '2018-04-01 02:13:55.123456 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2018-04-01 02:13:55.123456 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '2018-03-25 03:17:17.456789 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '2018-03-25 03:17:17.456789 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '1986-01-01 00:13:07.456789 UTC'",
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '1986-01-01 00:13:07.456789 UTC'")
+                .addRoundTrip(inputType, "TIMESTAMP '2021-09-07 23:59:59.999999 UTC'",
                         TIMESTAMP_TZ_MICROS, "TIMESTAMP '2021-09-07 23:59:59.999999 UTC'")
-                // max value in BigQuery
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:562]: Why did you remove max value test case? Please revert.

**Comment by regadas** [comment]
Hi @ebyhr can you take another look? thank you

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 18]
```diff
@@ -526,44 +533,125 @@ public void testTime()
...
+                        TIMESTAMP_TZ_MICROS, "TIMESTAMP '9999-12-31 23:59:59.999999 UTC'");
+    }
+
+    @DataProvider
+    public Object[][] sessionZonesDataProvider()
+    {
+        return new Object[][] {
+                {UTC},
+                {jvmZone},
+                // using two non-JVM zones so that we don't need to worry what Postgres system zone is
```
[the remark this answers is not in the document]
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java]: Ahah 🤦 done! Thx

**Comment by ebyhr** [comment]
/test-with-secrets sha=4fdb7eb72775cb8a3de06c0a243b78b9b6913a30

**Comment by ebyhr** [comment]
@regadas Please confirm the above CI failure. Some failure come from BigQuery SLA, but `TestBigQueryAvroConnectorTest.testDataMappingSmokeTest ` failure looks related to this change.

**Comment by regadas** [comment]
@ebyhr hm this seems a BQ related issue (probably due to resource availability for the project used). I'll rerun the tests on my side to confirm.

**Comment by regadas** [comment]
yup all good
```
[INFO] Tests run: 41, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 355.769 s - in io.trino.plugin.bigquery.TestBigQueryAvroTypeMapping
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 41, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:08 min (Wall Clock)
[INFO] Finished at: 2023-06-19T21:23:31-04:00
[INFO] ------------------------------------------------------------------------
[WARNING] 
[WARNING] Plugin validation issues were detected in 4 plugin(s)
[WARNING] 
[WARNING]  * com.hubspot.maven.plugins:dependency-scope-maven-plugin:0.10
[WARNING]  * org.jacoco:jacoco-maven-plugin:0.8.10
[WARNING]  * org.apache.maven.plugins:maven-dependency-plugin:3.5.0
[WARNING]  * org.codehaus.mojo:build-helper-maven-plugin:3.3.0
[WARNING] 
[WARNING] For more or less details, use 'maven.plugin.validation' property with one of the values (case insensitive): [BRIEF, DEFAULT, VERBOSE]
[WARNING] 
```

**Comment by ebyhr** [comment]
> this seems a BQ related issue

No, `testDataMappingSmokeTest` failure is related to this change.

**Comment by ebyhr** [comment]
> [INFO] Tests run: 41, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 355.769 s - in io.trino.plugin.bigquery.TestBigQueryAvroTypeMapping

The test is wrong. Please confirm `TestBigQueryAvroConnectorTest`.

**Comment by regadas** [comment]
>No, testDataMappingSmokeTest failure is related to this change.

unsure really because the logs do have 
```
Caused by: com.google.cloud.bigquery.BigQueryException: 502 Bad Gateway
```

**Comment by ebyhr** [comment]
Please take a look at:
```
Error:  Failures: 
Error:    TestBigQueryAvroConnectorTest>BaseConnectorTest.testDataMappingSmokeTest:5078->BaseConnectorTest.testDataMapping:5108->BaseConnectorTest.lambda$testDataMapping$67:5101->AbstractTestQueryFramework.assertUpdate:405->AbstractTestQueryFramework.assertUpdate:410 » QueryFailed An internal error occurred and the request could not be completed. This is usually caused by a transient issue. Retrying the job with back-off as described in the BigQuery SLA should solve the problem: https://cloud.google.com/bigquery/sla. If the error continues to occur please contact support at https://cloud.google.com/support.
Error:    TestBigQueryAvroConnectorTest>BaseConnectorTest.testDataMappingSmokeTest:5078->BaseConnectorTest.testDataMapping:5113->AbstractTestQueryFramework.assertQuery:339 For query 20230620_001638_01284_uyiad: 
 SELECT row_id FROM test_data_mapping_smoke_timestamp_6__with_time_zone5njyehw04u WHERE rand() = 42 OR value = TIMESTAMP '1969-12-31 15:03:00.123456 +01:00'
not equal
Actual rows (up to 100 of 0 extra rows shown, 0 rows in total):
    
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [sample value]

Error:    TestBigQueryAvroConnectorTest>BaseConnectorTest.testDataMappingSmokeTest:5078->BaseConnectorTest.testDataMapping:5113->AbstractTestQueryFramework.assertQuery:339 For query 20230620_001800_01293_uyiad: 
 SELECT row_id FROM test_data_mapping_smoke_timestamp_6__with_time_zone7n85c1xvdz WHERE rand() = 42 OR value = TIMESTAMP '2020-02-12 15:03:00 +01:00'
not equal
Actual rows (up to 100 of 0 extra rows shown, 0 rows in total):
    
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [sample value]
```

**Comment by regadas** [comment]
Will do! the `502` could be a red herring.

is there a faster way to run `testDataMappingSmokeTest`? the other one is still going

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 19]
```diff
@@ -265,7 +265,6 @@ protected Optional<DataMappingTestSetup> filterDataMappingSmokeTestData(DataMapp
             case "timestamp":
             case "timestamp(3)":
             case "timestamp(3) with time zone":
-            case "timestamp(6) with time zone":
```
Comment (regadas) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:268]: Ok go it so the tests fail because of this. Tracing the tests I do see that tests use invalid TIMESTAMP literals for BigQuery

https://github.com/trinodb/trino/blob/158f0b2f8748edf632a88f879d01ae208f08ed7d/testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java#L5169

Not really sure what do to here, add back this, or override @ebyhr?

Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:268]: It means you need to fix this PR, (NOT test).

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 20]
```diff
@@ -526,44 +533,125 @@ public void testTime()
...
+        testTimestampWithTimeZone("TIMESTAMP")
+                .execute(getQueryRunner(), bigqueryCreateAndInsert("test.timestamp_tz"));
+    }
+
+    private SqlDataTypeTest testTimestampWithTimeZone(String inputType)
+    {
+        return SqlDataTypeTest.create()
                 // min value in BigQuery
-                .addRoundTrip("TIMESTAMP", "TIMESTAMP '0001-01-01 00:00:00.000000 UTC'",
+                .addRoundTrip(inputType, "TIMESTAMP '0001-01-01 00:00:00.000 UTC'",
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:556]: This change loses timestamp literal with `Asia/Kathmandu` zone. Please don't reduce the test coverage.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 19 above)
Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:268]: Help me understand then a bit, there are no docs. isnt that literal `TIMESTAMP '2020-02-12 15:03:00 +01:00'` used in an BigQuery query?

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:268]: > It means you need to fix this PR, (NOT test)

TBF this is not really what I meant.

Comment (regadas) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:268]: addressed.

**Comment by regadas** [comment]
Hi @ebyhr can you take another look at this? Thank you 🙏

**Comment by ebyhr** [comment]
/test-with-secrets sha=e47f908d3ac06578266c586cff24d682512f1db3

**Comment by ebyhr** [comment]
It seems tests with secret didn't start correctly. I will re-trigger. 
cc: @nineinchnick

**Comment by ebyhr** [comment]
/test-with-secrets sha=e47f908d3ac06578266c586cff24d682512f1db3

**Comment by nineinchnick** [comment]
https://github.com/trinodb/trino/actions/runs/5476655936/jobs/9974512729#step:3:16
@wendigo permissions were not correct at that time.

**Comment by nineinchnick** [comment]
The workflow with secrets failed because there's an invalid condition, here's a fix: https://github.com/trinodb/trino/pull/18165

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java [hunk 21]
```diff
@@ -526,44 +533,129 @@ public void testTime()
...
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey(zoneId.getId()))
+                .build();
+
+        testTimestampWithTimeZone("TIMESTAMP(6) WITH TIME ZONE")
+                .execute(getQueryRunner(), trinoCreateAsSelect("test.timestamp_tz"))
+                .execute(getQueryRunner(), trinoCreateAsSelect(session, "test.timestamp_tz"))
+                .execute(getQueryRunner(), session, trinoCreateAsSelect("test.timestamp_tz"));
+
+        testTimestampWithTimeZone("TIMESTAMP")
+                .execute(getQueryRunner(), bigqueryCreateAndInsert("test.timestamp_tz"));
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryTypeMapping.java:549]: #12346
